graphql_compiler/tests/test_compiler.py

```diff
@@ -9662,10 +9677,14 @@ def test_simple_optional_recurse(self):
                 anon_1.[Animal_in_Animal_ParentOf__name] AS child_name,
                 anon_1.[Animal__name] AS name,
                 anon_2.name AS self_and_ancestor_name
-            FROM anon_1, anon_2
+            FROM
+                anon_1
+                LEFT OUTER JOIN anon_2
```

Comment:
I'm not sure why this is a `LEFT OUTER JOIN`. Isn't the `@optional` already handled by the `LEFT OUTER JOIN` on line 9651? As far as I can tell, because the recursive CTE always emits depth-0 recursions, this join should always have _something_ to join on, and `INNER JOIN` is therefore equivalent and probably semantically preferred.

Reply:
Yeah actually, that makes sense. I'll run the integration test to double-check. This will simplify the implementation (no need for `self._came_from` shenanigans)

Going back to the PR description, I'd need to think more about solution (a) vs (b). I think I'll continue with (b), but it's good to know that (a) is still on the table.

Reply:
If I understand correctly, I think joining the CTEs together is unavoidable because outputting from the recursive CTE vs the pre-recursion CTE might be semantically-different and corresponds to different input queries. Consider the following somewhat contrived but semantically-valid query:
```graphql
{
    Species {
        name @filter(op_name: "=", value: ["$dog"]) @output(out_name: "species")

        in_Animal_Species {
            name @output(out_name: "animal")

            out_Animal_ParentOf @recurse(depth: 1) {
                name @output(out_name: "self_or_child_animal")
            }
        }
    }
}
```

This is what I'd expect it to compile into for MSSQL:
```sql
WITH anon_1 AS (
    SELECT
        [Animal_1].name AS [Species_in_Animal_Species__name],
        [Animal_1].parent AS [Species_in_Animal_Species__parent],
        [Animal_1].uuid AS [Species_in_Animal_Species__uuid],
        [Species_1].name AS [Species__name]
    FROM
        db_1.schema_1.[Species] AS [Species_1]
        JOIN db_1.schema_1.[Animal] AS [Animal_1]
            ON [Species_1].uuid = [Animal_1].species
    WHERE
        [Species_1].name = :dog
),
anon_2(name, parent, uuid, __cte_key, __cte_depth) AS (
    SELECT
        anon_1.[Species_in_Animal_Species__name] AS name,
        anon_1.[Species_in_Animal_Species__parent] AS parent,
        anon_1.[Species_in_Animal_Species__uuid] AS uuid,
        anon_1.[Species_in_Animal_Species__uuid] AS __cte_key,
        0 AS __cte_depth
    FROM
        anon_1
    UNION ALL
    SELECT
        [Animal_2].name AS name,
        [Animal_2].parent AS parent,
        [Animal_2].uuid AS uuid,
        anon_2.__cte_key AS __cte_key,
        anon_2.__cte_depth + 1 AS __cte_depth
    FROM
        anon_2
        JOIN db_1.schema_1.[Animal] AS [Animal_2]
            ON anon_2.parent = [Animal_2].uuid
        WHERE anon_2.__cte_depth < 2
)
SELECT
    anon_1.[Species_in_Animal_Species__name] AS animal,
    anon_2.name AS self_or_child_animal,
    anon_1.[Species__name] AS species
FROM
    anon_1
    JOIN anon_2
        ON anon_1.[Species_in_Animal_Species__uuid] = anon_2.__cte_key
```

I don't see where we have the option to pick whether we output from `anon_1` or `anon_2`. Easy cases first:
- The value for `species` can only come from `anon_1`, hopefully that's self-evident since it has nothing to do with the recursion.
- The value for `self_or_child_animal` has to come from `anon_2` because `anon_1` doesn't know about it.

The only somewhat tricky case is the value for `animal`. I think we can rule out `anon_2.name` as the source for this value, because that's clearly the source for `self_or_child_animal` which has different semantics per the `@recurse`. `anon_2` doesn't output anything else on the basis of a `name` column (the explicitly-declared options are `name, parent, uuid, __cte_key, __cte_depth`), so I think this value has to come from `anon_1.[Species_in_Animal_Species__name]`.

So I've either entirely missed your point (and if so, I apologize), or it seems that the outputs from outside the `@recurse` block _must_ come from `anon_1` and the `JOIN` between the two CTEs is unavoidable.

Reply:
`species` can come from `anon_2` if `anon_2` selects it in the base case, and copies it in the recursive step.

Reply:
> `species` can come from `anon_2` if `anon_2` selects it in the base case, and copies it in the recursive step.

Sure — but that seems unnecessary. We'd just be carrying a known-constant value (from the perspective of any input row to `anon_2`) through the entire recursion without ever computing anything off of it. This would just add cost to the database computation, so I'd rather avoid it unless the code gets too gnarly.

> Back to the first comment in the chain: It's actually necessary to `LEFT JOIN`. On `@optional` miss, the preceding selectable will have a `NULL` value in the id column that joins to the recursive cte. So Even though the recursive CTE base case contains the whole of the preceding selectable, you can't join on `NULL`.

Ah yes, because of the ordering of steps (`anon_1` before `anon_2`) that's true. I forgot that the key isn't just coming from a `FROM Animal` clause and could therefore be `null`. Good catch!